### PR TITLE
Feat/rbf decrease

### PR DESCRIPTION
**Pull request by szymonlesisz**
depends on https://github.com/trezor/connect/pull/815

To complete the RBF experience following scenarios are added:

- transaction with OP_RETURN output
- transaction without change output (set-max)
- possibility to add another utxo (input) to existing transaction
- possibility to decrease output of existing transaction

# QA testing scenarios:
[temporary testing link](https://suite.corp.sldev.cz/suite-web/feat/rbf-decrease/web/?trezor-connect-src=https://connect.corp.sldev.cz/feat/rbf-opreturn/)
Except for current possible scenarios (decreasing change output) you can:
1.
- send ALL from the account to one external output
- bump fee by decreasing sent amount
Screenshot from 2021-06-02 17-00-49

2.
- send ALL from the account to multiple external outputs
- bump fee by decreasing amount of the output of your choice
Screenshot from 2021-06-02 17-00-29

3.
- send ALL from the account
- then receive some utxo to that account and use it to bump fee 

4.
- bump fee of transaction with OP_RETURN output 
(you can combine it with steps 1-3: one OP_RETURN output + one or more external outputs + set-max)

## Notes
- "native" RBF is available since FW 1.9.4/2.3.5
- "native" decrease output is available since FW 1.10.0/2.4.0
- in case where native RBF/decrease is NOT available there should be a fallback to "legacy" way of signing (just regular signing, like from send form)

### TT
Screenshot from 2021-06-02 17-40-33
Screenshot from 2021-06-02 17-39-16
-------
### T1
Screenshot from 2021-06-02 17-47-58
Screenshot from 2021-06-02 17-48-13

### CODE REVIEW

File: packages/suite/src/utils/wallet/transactionUtils.ts
Comment (goodhoko) [inline packages/suite/src/utils/wallet/transactionUtils.ts]: What does `6a - OP_RETURN and XX - data len` mean? I don't get it. 🤷

**Review by goodhoko — APPROVED** [review]
Just one small question. Otherwise seems good to me. .)

File: packages/suite/src/utils/wallet/transactionUtils.ts
Comment (szymonlesisz) [reply packages/suite/src/utils/wallet/transactionUtils.ts]: https://en.bitcoin.it/wiki/Script#Constants

blockbook returns hex data prefixed with 2 constants and we want data without it.

example:
https://tbtc1.trezor.io/api/v2/tx/6039e108e1dfc532c3ae9eceaf6f7e87e967083ea23ad2d7444383ce1ab50eca

data returned by BB in output.hex `6a03666f6f`
data retuned by BB in output.addresses![0] field: `OP_RETURN (foo)`
we want: `666f6f` (ascii foo)

unwanted prefix:
6a - OP_RETURN constant
03 - data length

Comment (goodhoko) [reply packages/suite/src/utils/wallet/transactionUtils.ts]: Aha! Thanks for the link to bitcoin wiki. I finally learned what the BTC scripts are. 🙈 

So if I understand this correctly `output.hex` and `output.addresses[0]` always hold the same data only differently encoded. Here we *only* use the later. That might be what confused me. Why are we mentioning the `hex` field when it's not used in the code?

Maybe we could change the comment like so:
```
             // TODO: this should be done in trezor-connect, blockchain-link or even blockbook
             // Extract data from BlockBook format which takes form of `OP_RETURN (<data in ascii>)`.
```

I.e. completely omit the part about `output.hex`?

However, this is rather minor thing. So feel free to RBY. .)

**Comment by goodhoko** [comment]
This is good to go from my side. The one open thread can be resolved by Szymon. Let's merge this! .)

File: packages/suite/src/utils/wallet/transactionUtils.ts
Comment (szymonlesisz) [reply packages/suite/src/utils/wallet/transactionUtils.ts]: `Why are we mentioning the hex field when it's not used in the code?`

because `hex` should be used instead of `address[0]` which is totally confusing.
valid solution is to use `@trezor/utxo-lib` like [in connect](https://github.com/trezor/connect/blob/develop/src/js/core/methods/tx/refTx.js#L115) but is not available here and adding it as a dependency only for this operation (minor use case) is unnecessary.

that is why i've commented that it should be fixed on totally different level (connect/blockchain-link)

i'll update the comment as you suggested

**Comment by szymonlesisz** [comment]
> Let's merge this

no, wait, i need to publish connect first, then i'm waiting for response from @matejzak on slack (suite-ui), and then we will merge it.

**Comment by goodhoko** [comment]
> no, wait, i need to publish connect first, then i'm waiting for response from @matejzak on slack (suite-ui), and then we will merge it.

Sure. Sorry for confusion. I mean: merge this when you're ready. .)

**Comment by szymonlesisz** [comment]
everything is rebased now, so i guess its ready to merge

**Comment by bosomt** [comment]
I was able to bump transaction multiple times,
But ended up with negative amount in source account :)
Otherwise went well and and I will test it multiple times during new release cycle.


image
image

image

https://btc1.trezor.io/tx/8af45dd4fd79e0c975bf934813917905c2ba8311dc2274fd51ee683e2e6c3f97

**Info:**
 - Suite version: web 21.6.2 (97d032adf84e5b7c5dd82fb3085fdeb4434ad7f8)
 - Browser: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:89.0) Gecko/20100101 Firefox/89.0
 - OS: Linux x86_64
 - Screen: 2560x1440
 - Device: model One 1.10.1 regular

**Comment by szymonlesisz** [comment]
i'm aware of negative balance bug, it should be visible only for a while, no? after few seconds it goes back to 0

 the fix is on the way

**Comment by bosomt** [comment]
@szymonlesisz i don't know for how long but went to correct values now and transaction is already confirmed

**Comment by szymonlesisz** [comment]
@bosomt negative balance should be fixed now

**Comment by bosomt** [comment]
@szymonlesisz yes I can confirm that it was fixed and works as expected in 4e29f3d6e527618670936bf6d1d3617b217be30c


one more thing, do you see that `$4.80`? it's not fee, it's amount that we are sending, but at that place looks like fee to me , is that expected ? i would expect previous fee at that position 


image

image


image

**Comment by bosomt** [comment]
**QA OK**


**Info:**
 - Suite version: desktop 21.7.0 (6796f451738b4cd59ecd9efb3257743bbe8398a1)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 11_4_0) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuite/21.7.0 Chrome/89.0.4389.69 Electron/12.0.0 Safari/537.36
 - OS: MacIntel
 - Screen: 1440x900
 - Device: model T 2.4.0 regular

**Pull request merged**
